### PR TITLE
[x/act V2]: Add MsgVote handling

### PR DESCRIPTION
**Pull request by artur-abliazimov**
https://github.com/orgs/warden-protocol/projects/12/views/1?pane=issue&itemId=78574759

## Summary by CodeRabbit

- **New Features**
	- Introduced enhanced voting logic for actions, allowing for better handling of approved and rejected votes.
	- Added functionality to create and manage action votes, including checks for action status and duplicate votes.
	- Implemented error handling for existing votes to improve robustness.

- **Bug Fixes**
	- Resolved issues related to action timeout handling in the voting process.

### CODE REVIEW

File: warden/x/act/keeper/actions.go [hunk 1]
```diff
@@ -30,6 +30,73 @@ func (approvers ApproversEnv) Get(name string) (object.Object, bool) {
 
 var _ shield.Environment = ApproversEnv{}
```
Comment (backsapc) [inline warden/x/act/keeper/actions.go]: @Pitasi why do we need this here?

File: warden/x/act/types/v1beta1/action.go [hunk 2]
```diff
@@ -15,6 +15,14 @@ func NewApprover(address string, timestamp time.Time) *Approver {
 	}
 }
 
+func NewVote(participant string, vote ActionVoteType, timestamp time.Time) *ActionVote {
+	return &ActionVote{
+		Participant: participant,
+		VotedAt:     timestamp,
+		Vote:        vote,
```
Comment (backsapc) [inline warden/x/act/types/v1beta1/action.go]: VoteType?

File: warden/x/act/types/v1beta1/errors.go [hunk 3]
```diff
@@ -23,4 +23,5 @@ var (
 	ErrInvalidRuleDefinition     = sdkerrors.Register(ModuleName, 1112, "invalid rule definition")
 	ErrInvalidRevoker            = sdkerrors.Register(ModuleName, 1113, "this account can't revoke this action")
 	ErrInvalidUpdateRuleAccount  = sdkerrors.Register(ModuleName, 1114, "this account can't update this rule")
+	ErrActionVoteAlreadyExists   = sdkerrors.Register(ModuleName, 1115, "vote already exists")
```
Comment (backsapc) [inline warden/x/act/types/v1beta1/errors.go]: User already voted?

**Review by backsapc — DISMISSED** [review]

File: warden/x/act/keeper/actions.go (hunk 1 above)
Comment (mn13) [reply warden/x/act/keeper/actions.go]: I guess it's a type check.

File: warden/x/act/keeper/actions.go [hunk 4]
```diff
@@ -28,7 +28,70 @@ func (approvers ApproversEnv) Get(name string) (object.Object, bool) {
...
+func (votes ActionRejectedVotesEnv) Get(name string) (object.Object, bool) {
+	for _, s := range votes {
+		if s.Participant == name && s.VoteType == types.ActionVoteType_VOTE_TYPE_REJECTED {
+			return object.TRUE, true
+		}
+	}
+	return object.FALSE, true
+}
+
+// TryExecuteVotedAction checks if the action's intent is satisfied and stores the
```
Comment (mn13) [inline warden/x/act/keeper/actions.go]:
```suggestion
// TryExecuteVotedAction checks if the action's expression is satisfied and stores the
```

**Review by backsapc — DISMISSED** [review]

**Review by mn13 — CHANGES_REQUESTED** [review]
I believe that AddApprover calling in AddAction (v1beta1/action.go 243) should be replaced with AddVote (where we add initial approver for action)

**Comment by artur-abliazimov** [comment]
> I believe that AddApprover calling in AddAction (v1beta1/action.go 243) should be replaced with AddVote (where we add initial approver for action)

I do not think that the initial approver must be set during action creation. AddVote is added separately `msg_server_vote_for_action.go` => `VoteForAction`. 

@Pitasi what do you think?

File: warden/x/act/keeper/actions.go [hunk 5]
```diff
@@ -29,7 +29,70 @@ func (approvers ApproversEnv) Get(name string) (object.Object, bool) {
...
+		if s.Participant == name && s.VoteType == types.ActionVoteType_VOTE_TYPE_REJECTED {
+			return object.TRUE, true
+		}
+	}
+	return object.FALSE, true
+}
+
+// TryExecuteVotedAction checks if the action's expression is satisfied and stores the
+// result in the database.
+func (k Keeper) TryExecuteVotedAction(ctx context.Context, act *types.Action) error {
```
Comment (pitasi) [inline warden/x/act/keeper/actions.go:59]: I'd like to see this split into two:

- TryExecuteAction (called when someone adds an approval)
- TryRejectAction (called when someone adds a rejection)

wdyt?

**Review by pitasi — CHANGES_REQUESTED** [review]

File: warden/x/act/keeper/actions.go (hunk 5 above)
Comment (artur-abliazimov) [reply warden/x/act/keeper/actions.go:59]: Yes, this makes sense. Thanks!

File: warden/x/act/keeper/actions.go [hunk 6]
```diff
@@ -29,7 +28,70 @@ func (approvers ApproversEnv) Get(name string) (object.Object, bool) {
...
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+
+	rejected, err := act.Rule.Eval(ctx, ActionRejectedVotesEnv(act.Votes))
+
+	if err != nil {
+		return err
+	}
+
+	if rejected {
+		if err := act.SetStatus(sdkCtx, types.ActionStatus_ACTION_STATUS_REVOKED); err != nil {
```
Comment (backsapc) [inline warden/x/act/keeper/actions.go:85]: Do we keep `Revoked` action status here?

Comment (backsapc) [reply warden/x/act/keeper/actions.go:85]: Do we need to add `Rejected`?

Comment (artur-abliazimov) [reply warden/x/act/keeper/actions.go:85]: I think that action can be either completed or revoked/timeout. Approve/Reject is for votes.

**Comment by pitasi** [comment]
> > I believe that AddApprover calling in AddAction (v1beta1/action.go 243) should be replaced with AddVote (where we add initial approver for action)
> 
> I do not think that the initial approver must be set during action creation. AddVote is added separately `msg_server_vote_for_action.go` => `VoteForAction`. 
> 
> @Pitasi what do you think?

i think that 99% of the times if I'm creating an action I also want to approve it, so it makes sense to save some time and gas and do it

File: warden/x/act/keeper/actions.go (hunk 6 above)
Comment (pitasi) [reply warden/x/act/keeper/actions.go:85]: if we want we can rename REVOKED to REJECTED for clarity

File: warden/x/act/types/v1beta1/action.go [hunk 7]
```diff
@@ -63,3 +71,28 @@ func (a *Action) AddApprover(ctx sdk.Context, address string) error {
...
 }
+
+func (a *Action) AddVote(ctx sdk.Context, participant string, voteType ActionVoteType) error {
+	if a.Status != ActionStatus_ACTION_STATUS_PENDING {
+		return errors.Wrapf(ErrInvalidActionStatus, "can't add a vote to an action that's not pending")
+	}
+
+	for _, v := range a.Votes {
+		if v.Participant == participant {
+			return ErrAlreadyParticipatedInVoting
```
Comment (pitasi) [inline warden/x/act/types/v1beta1/action.go]: i think we should also allow an user to change their mind, i don't see any particular reason for making the vote immutable

File: warden/x/act/types/v1beta1/action.go [hunk 8]
```diff
@@ -63,3 +71,35 @@ func (a *Action) AddApprover(ctx sdk.Context, address string) error {
...
+func (a *Action) AddOrUpdateVote(ctx sdk.Context, participant string, voteType ActionVoteType) error {
+	if a.Status != ActionStatus_ACTION_STATUS_PENDING {
+		return errors.Wrapf(ErrInvalidActionStatus, "can't add a vote to an action that's not pending")
+	}
+
+	updated := false
+
+	for i, v := range a.Votes {
+		if v.Participant == participant {
+			a.Votes[i].VoteType = voteType
```
Comment (pitasi) [inline warden/x/act/types/v1beta1/action.go:84]:
```suggestion
			a.Votes[i].VoteType = voteType
			a.Votes[i].VotedAt = ctx.BlockTime()
```

**Review by pitasi — CHANGES_REQUESTED** [review]

File: warden/x/act/keeper/msg_server_vote_for_action.go [hunk 9]
```diff
@@ -2,10 +2,49 @@ package keeper
...
+
+	if msg.VoteType == types.ActionVoteType_VOTE_TYPE_APPROVED {
+		if err := k.TryExecuteVotedAction(ctx, &act); err != nil {
+			return nil, err
+		}
+	} else if msg.VoteType == types.ActionVoteType_VOTE_TYPE_REJECTED {
+		if err := k.TryRejectVotedAction(ctx, &act); err != nil {
+			return nil, err
+		}
+	}
```
Comment (pitasi) [inline warden/x/act/keeper/msg_server_vote_for_action.go:50]: who knows what's going to happen in the future, let's reject unhandled values:

```suggestion
	} else {
		return nil, fmt.Errorf("unhandled VoteType value: %v", msg.VoteType)
	}
```

**Review by pitasi — CHANGES_REQUESTED** [review]
one last thing then looks good 👌

**Review by pitasi — APPROVED** [review]

**Review by mn13 — APPROVED** [review]

**Pull request merged**